### PR TITLE
[KYUUBI #1516] Implement api: /${version}/operations/${operation_identifier}

### PR DESCRIPTION
**Pull request by simon824**
### _Why are the changes needed?_
This is a subtask of umbrella issue #KPIP-1

/${version}/operations/${operation_identifier}
- desc: apply an action for an operation based on a given session identifier and operation identifier
- method: PUT
- params:
- action: "CANCEL" or "ADD" or "CLOSE"
- returns: status code

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by simon824** [comment]
> * action: "CANCEL" or "ADD" or "CLOSE"

"ADD" action maybe is unnecessary here, when we create an operation ,  `addOperation` will always be called.  WDYT? @yanghua

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/OperationsResource.scala [hunk 1]
```diff
@@ -52,4 +53,31 @@ private[v1] class OperationsResource extends ApiRequestContext {
...
+  @PUT
+  @Path("{operationHandle}")
+  def applyOpAction(
+      request: OpActionRequest,
+      @PathParam("operationHandle") operationHandleStr: String): Response = {
+    try {
+      val operationHandle = parseOperationHandle(operationHandleStr)
+      val operationManager = backendService.sessionManager.operationManager
+
+      request.action.toLowerCase() match {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/OperationsResource.scala:72]: backendService.close/cancelOperation

**Comment by yanghua** [comment]
> 

sounds good, reasonable

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/OperationsResource.scala [hunk 2]
```diff
@@ -52,4 +53,31 @@ private[v1] class OperationsResource extends ApiRequestContext {
...
+        case "cancel" => operationManager.cancelOperation(operationHandle)
+        case "close" => operationManager.closeOperation(operationHandle)
+        case _ => throw KyuubiSQLException(s"Invalid request $request")
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/OperationsResource.scala:75]: Invalid action looks better?

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/OperationsResourceSuite.scala [hunk 3]
```diff
@@ -17,56 +17,85 @@
...
+        .request(MediaType.APPLICATION_JSON_TYPE)
+        .put(Entity.entity(OpActionRequest("close"), MediaType.APPLICATION_JSON_TYPE))
       assert(200 == response.getStatus)
-      assert(operationDetail.operationStatus.state == OperationState.FINISHED)
 
-      // Invalid operationHandleStr
-      val invalidOperationHandle = s"${operationHandle.identifier.publicId}|" +
-        s"${operationHandle.identifier.secretId}|${operationHandle.protocol.getValue}|GET_TYPE_INFO"
-      response = webTarget.path(s"api/v1/operations/$invalidOperationHandle")
+      response = webTarget.path(s"api/v1/operations/$opHandleStr")
```
Comment (yanghua) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/OperationsResourceSuite.scala:68]: also check the `OperationState.CLOSED`?

Comment (simon824) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/OperationsResourceSuite.scala:68]: Once the operation is closed, the `Object Operation` cannot be obtained, so cannot get `OperationState` too.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/OperationsResource.scala [hunk 4]
```diff
@@ -52,4 +53,29 @@ private[v1] class OperationsResource extends ApiRequestContext {
...
+  @Path("{operationHandle}")
+  def applyOpAction(
+      request: OpActionRequest,
+      @PathParam("operationHandle") operationHandleStr: String): Response = {
+    try {
+      val operationHandle = parseOperationHandle(operationHandleStr)
+      request.action.toLowerCase() match {
+        case "cancel" => backendService.cancelOperation(operationHandle)
+        case "close" => backendService.closeOperation(operationHandle)
+        case _ => throw KyuubiSQLException(s"Invalid action $request")
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/OperationsResource.scala:73]: `$request` -> `${request.action}`?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/OperationsResource.scala [hunk 5]
```diff
@@ -52,4 +53,29 @@ private[v1] class OperationsResource extends ApiRequestContext {
...
+      }
+      Response.ok().build()
+    } catch {
+      case NonFatal(_) =>
+        throw new NotFoundException(s"Error applying ${request.action} for $operationHandleStr")
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/OperationsResource.scala:78]: `for operation handle $operationHandleStr `?

**Review by yanghua — COMMENTED** [review]
Left two minor comments, otherwise LGTM.

**Review by yanghua — APPROVED** [review]
LGTM

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/dto.scala [hunk 6]
```diff
@@ -83,3 +83,5 @@ case class OperationDetail(
     shouldRunAsync: Boolean,
     isTimedOut: Boolean,
     operationStatus: OperationStatus)
+
+case class OpActionRequest(action: String)
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/dto.scala:87]: How about CancelOrCloseOperationRequest？

Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/dto.scala:87]: or maybe we divide them into 2 apis?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/OperationsResource.scala [hunk 7]
```diff
@@ -52,4 +53,30 @@ private[v1] class OperationsResource extends ApiRequestContext {
...
     }
   }
+
+  @ApiResponse(
+    responseCode = "200",
+    content = Array(new Content(
+      mediaType = MediaType.APPLICATION_JSON)),
+    description =
+      "apply an action for an operation")
+  @PUT
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/OperationsResource.scala:63]: shall it be POST, i checked at Livy's API design, cc @yanghua

**Pull request closed without merging**